Document/0x06h-Testing-Platform-Interaction.md

```diff
@@ -3,365 +3,2701 @@
+
+```
+// __handlers__/TelegramUI/_S10TelegramUI15openExternalUrl7_b1a3234e.js
+
+  onEnter: function (log, args, state) {
+
+    log("TelegramUI.openExternalUrl(account: TelegramCore.Account, 
+        context: TelegramUI.OpenURLContext, url: Swift.String, forceExternal: Swift.Bool, 
+        presentationData: TelegramUI.PresentationData, 
+        applicationContext: TelegramUI.TelegramApplicationContext, 
+        navigationController: Display.NavigationController?, dismissInput: () -> ()) -> ()");
+    log("\taccount: " + ObjC.Object(args[0]).toString());
+    log("\tcontext: " + ObjC.Object(args[1]).toString());
+    log("\turl: " + ObjC.Object(args[2]).toString());
+    log("\tpresentationData: " + args[3]);
+    log("\tapplicationContext: " + ObjC.Object(args[4]).toString());
+    log("\tnavigationController: " + ObjC.Object(args[5]).toString());
+  },
+```
+
+This way, the next time we run it we get an much more detailed output:
+
+```
+298382 ms  -[AppDelegate application:0x10556b3c0 continueUserActivity:0x1c4237780 
+                restorationHandler:0x16f27a898]
+298382 ms  	application:<Application: 0x10556b3c0>
+298382 ms  	continueUserActivity:<NSUserActivity: 0x1c4237780>
+298382 ms  		webpageURL:http://t.me/addstickers/radare
+298382 ms  		activityType:NSUserActivityTypeBrowsingWeb
+298382 ms  		userInfo:{
+}
+298382 ms  	restorationHandler:<__NSStackBlock__: 0x16f27a898>
+
+298619 ms     | TelegramUI.openExternalUrl(account: TelegramCore.Account, 
+context: TelegramUI.OpenURLContext, url: Swift.String, forceExternal: Swift.Bool, 
+presentationData: TelegramUI.PresentationData, applicationContext: 
+TelegramUI.TelegramApplicationContext, navigationController: Display.NavigationController?, 
+dismissInput: () -> ()) -> ()
+298619 ms     | 	account: TelegramCore.Account
+298619 ms     | 	context: nil
+298619 ms     | 	url: http://t.me/addstickers/radare
+298619 ms     | 	presentationData: 0x1c4e40fd1
+298619 ms     | 	applicationContext: nil
+298619 ms     | 	navigationController: TelegramUI.PresentationData
+```
+
+There you can observe the following:
+
+- it calls `application:continueUserActivity:restorationHandler:` from the app delegate as expected
+- `application:continueUserActivity:restorationHandler:` handles the URL but does not open it, it calls `TelegramUI.openExternalUrl` for that
+- the URL being opened is `https://t.me/addstickers/radare`
+
+You can now keep going and try to trace and verify how the data is being validated. For example, if you have two apps that *communicate* via universal links you can use this to see if the sending app is leaking sensitive data by hooking these methods in the receiving app. This is especially useful when you don't have the source code as you will be able to retrieve the full URL that you wouldn't see other way as it might be the result of clicking some button or triggering some functionality.
```

Comment:
I'm not sure if I understand the question. Are you referring to section "Checking if the App is Calling Other App's Universal Links" of the static analysis? 

What do you mean with *hook sensitive methods* and *extract data*? I could hook virtually anything :P I just need to know what. For that, if I don't have the code I would do exactly what's described in this section and keep iterating on the traces until I find all methods and relevant data. Should I rephrase something to express this better?

Reply:
Basically: how do you make sure that the receiving app is hardened enough to handle all the types of input it gets: so did we make sure that the app that handles the incoming link is hardened against msiuse (e.g. no authentication bypass, information leakage, etc.).

Reply:
I would say this is covered already in the **static analysis** sections:
- Checking the Link Receiver Method
- Checking the Data Handler Method (mostly here)

If needed we can open a new issue in case it is not yet clear enough. 

Here's an excerpt:
___

###### Checking the Data Handler Method

You should check how the received data is validated. Apple [explicitly warns about this](https://developer.apple.com/documentation/uikit/core_app/allowing_apps_and_websites_to_link_to_your_content/handling_universal_links "Handling Universal Links"):

> Universal links offer a potential attack vector into your app, so make sure to validate all URL parameters and discard any malformed URLs. In addition, limit the available actions to those that do not risk the user’s data. For example, do not allow universal links to directly delete content or access sensitive information about the user. When testing your URL-handling code, make sure your test cases include improperly formatted URLs.

[...]

In addition, remember that if the URL includes parameters, they should not be trusted before being carefully sanitized and validated (even when including a whitelist of trusted domains here). For example, they might have been spoofed by an attacker or might include malformed data. If that is the case, the whole URL and therefore the universal link request must be discarded.

[...]

Finally, as stated above, be sure to verify that the actions triggered by the URL do not expose sensitive information or risk the user’s data on any way.

___

Here, in the **dynamic analysis** (where this comment refers), we have the following:

___

###### Checking How the Links Are Opened

If you want to know more about which function actually opens the URL and how the data is actually being handled you should keep investigating.

Extend the previous command in order to find out if there are any other functions involved into opening the URL.

[...]

You can now keep going and try to trace and verify how the data is being validated. For example, if you have two apps that *communicate* via universal links you can use this to see if the sending app is leaking sensitive data by hooking these methods in the receiving app. This is especially useful when you don't have the source code as you will be able to retrieve the full URL that you wouldn't see other way as it might be the result of clicking some button or triggering some functionality.
___